File: alura/poo/bytebank-herdado-conta/src/Conta.java
public abstract class Conta {
  protected double saldo;
  private int agencia;
  private int numero;
  private Cliente titular;
  private static int total;

  public Conta() {

  }

  public Conta(int numero, int agencia) {
    Conta.total++;
    setAgencia(agencia);
    setNumero(numero);
  }

  public abstract void deposita(double valor);

  public void saca(double valor) {
    if(this.saldo < valor) {
      throw new SaldoInsuficienteException("Saldo : " + getSaldo() + " Valor: " + valor);
    }

    this.saldo -= valor;
  }

  public void transfere(double valor, Conta destino) {
      this.saca(valor);
      destino.deposita(valor);
  }

  public double getSaldo() {
    return this.saldo;
  }

  public int getNumero() {
    return this.numero;
  }

  public void setNumero(int novoNumero) {
    if (novoNumero <= 0) {
      System.out.println("Número Inválido!");
      return;
    }
    this.numero = novoNumero;
  }

  public int getAgencia() {
    return agencia;
  }

  public void setAgencia(int agencia) {
    if(agencia <= 0) {
      System.out.println("Agencia inválida!");
      return;
    }
    this.agencia = agencia;
  }

  public void setTitular(Cliente titular) {
    this.titular = titular;
  }

  public Cliente getTitular() {
    return titular;
  }

  public static int getTotal() {
    return total;
  }
}

File: alura/Alura-Kotlin/kotlin/kotlin-orientacao-objetos/untitled/src/main/kotlin/Conta.kt
abstract class Conta(
  val titular: String,
  val numero: Int) {

  var saldo = 0.0
    private set;


  fun deposita(valor: Double) {
    this.saldo += valor;
  }

  open fun sacar(valor: Double): Boolean {
    if (saldo - valor >= 0.0) {
      saldo -= valor;
      return true;
    }

    return false;
  }

  fun transfere(valor: Double, destino: Conta): Boolean {
    if (saldo >= valor) {
      saldo -= valor;
      destino.deposita(valor);
      return true;
    }

    return false;
  }


}